python/nav/eventengine/export.py

```diff
@@ -0,0 +1,88 @@
+#
+# Copyright (C) 2019 UNINETT
+#
+# This file is part of Network Administration Visualized (NAV).
+#
+# NAV is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License version 3 as published by
+# the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.  You should have received a copy of the GNU General Public License
+# along with NAV. If not, see <http://www.gnu.org/licenses/>.
+#
+"""Alert stream export functionality"""
+import logging
+import json
+
+from django.core.serializers.json import DjangoJSONEncoder
+
+try:
+    import subprocess32 as subprocess
+except ImportError:
+    import subprocess
+from django.utils import six
+
+from nav.web.api.v1.alert_serializers import AlertQueueSerializer
+
+try:
+    exporter
+except NameError:
+    """A singleton, normally instantiated by the event engine as it starts"""
+    exporter = None
+
+_logger = logging.getLogger(__name__)
+
+
+class StreamExporter(object):
+    """Exports a stream of alert objects over a pipe to a specific script"""
+
+    def __init__(self, command):
+        self.command = command
+        self._process = None
+        self.run()
+
+    def run(self):
+        """Runs the subprocess that will receive the alert stream on its STDIN"""
+        self._process = subprocess.Popen(
+            [self.command], stdin=subprocess.PIPE, stdout=subprocess.DEVNULL
+        )
+
+    def is_ok(self):
+        """Ensures the subprocess is running.
+
+        :returns: True if the subprocess is running or could be restarted
+                  successfully. Otherwise, False.
+
+        """
+        while self._process.poll() is not None:
+            _logger.info(
+                "restarting dead export script (retcode=%s) %r",
+                self._process.returncode,
+                self.command,
+            )
+            self.run()
+
+        return True
```

Comment:
Doesn't look like `is_ok` can ever return anything Falsey?

Reply:
Not at the moment, no. It needs more error handling logic, as this implementation may get stuck in a loop trying to start a command that will never work. It needs to give up at some point and return `False` instead.